text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,655 @@
+got to keep writing down the information in so many places!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Cargo's manifest parsing will be updated with new features to support
+deduplicating each of the areas above. While all of these new features are
+pretty small in their own right, they all add up to greatly reducing the
+overhead of managing a workspace of many crates. The list of new features in
+Cargo will look like the following:
+
+## Workspace-level Dependencies
+
+The `[workspace]` section can now have a `dependencies` section which works the
+same way as the `[dependencies]` section in `Cargo.toml`:
+
+```toml
+# in workspace's Cargo.toml
+[workspace.dependencies]
+log = "0.3.1"
+log2 = { version = "2.0.0", package = "log" }
+serde = { git = 'https://github.com/serde-rs/serde' }
+wasm-bindgen-cli = { path = "crates/cli" }
+```
+
+Each workspace member can then reference this section in the workspace with a
+new dependency directive:
+
+```toml
+# in a workspace member's Cargo.toml
+[dependencies]
+log = { workspace = true }
+```
+
+This directive indicates that the `log` dependency should be looked up from
+`workspace.dependencies` in the workspace root. You can reference any name
+defined in `[workspace.dependencies]` too:
+
+```toml
+[dependencies]
+log2 = { workspace = true }
+```
+
+## No longer need both `version` and `path` to publish to Crates.io
+
+When you have a `path` dependency, Cargo's current behavior on publication looks
+like this:
+
+* If you have a `version` specifier as well, then the `path` key is deleted and
+  the crate is uploaded with the specified `version` as a dependency
+  requirement.
+* If you don't have a `version` specifier, then the dependency directive is
+  deleted and crates.io will not learn about this dependency. This is only
+  really useful for `dev-dependencies`.
+
+Cargo's behavior will change in this second case, instead following new logic
+for a missing `version` specifier. For dev-dependencies where the referenced
+package is `publish = false`, then the dependency will be dropped. Otherwise
+Cargo will assume that `version = "$dependency_version"` was specified, meaning
+that it requires at least the current version and otherwise any
+semver-compatible version.
+
+This behavior should mean that you no longer need to write `version = "..."`
+with `path` dependencies if you publish to crates.io. Coupled with the
+workspace-level dependencies above this means you never have to write the
+version of a path dependency anywhere!
+
+## Package metadata can reference other workspace members
+
+To deduplicate `[package]` directives in `Cargo.toml` workspace members, Cargo
+will now support declaring that metadata directives should be inherited from the
+workspace. For example to version every package the same within a workspace you
+can specify:
+
+```toml
+[package]
+name = "foo"
+version = { workspace = true }
```

Comment:
I left a comment earlier on using `from_workspace` instead of `workspace`, I don't know why but did not get enough attention. Please can you consider it, since it is syntactically more accurate and easier to understand. `workspace` sound like a feature flag.

Reply:
@snnsnn Can we even bikeshed on this? I wonder if there are any better alternatives without using `_` or just one word while still being easy to understand.

Reply:
How about `version = "@workspace"`?

Reply:
I wish it could be `version = "workspace"` haha.

Reply:
@pickfire `import`, maybe? But still, if `from_workspace` is too long, it is a problem because it's repeated at every place. ;)

Reply:
Or like what @pksunkara mentioned `foo = "workspace"` or what @alexcrichton mentioned `foo = "ws"` but theirs describe it on the package level.

Reply:
"ws" resembles websocket too much IMO.

Reply:
Some suggestions I find self-explanatory and not too long:
1. `from = "workspace"`
2. `src = "workspace"`
3. `import = "workspace"`
4. `from = "parent"` (if there's always a tree-like hierarchy)
(other permutations)

Reply:
It is not like version number you constantly change, so there is no real benefit from using abbreviations other than saving few keystrokes, as noted above it leads to more confusion. 

Explicitness and coherence should be the ones we strive for because whatever you put in there, will be there forever, well almost.

I checked [cargo's manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html) for convention or similarity, kebab-case is used for multi-word keys, so `from-workspace = true` would be better choice if you decide to go with a boolean. However string values works equally well and it provides more flexibility if there will be multiple sources to derive from beside the workspace, assuming there is verification mechanism to enforce a valid value. In that case `from = "workspace"` or `inherit-from = "workspace"` sounds fine since both of them clearly indicate value is derived. But `from` itself is enough, since there is From trait, there will be logical proximity or mental association, so it won't be that much confusing or vague.